pkg/webhook/preflight/nutanix/failuredomain.go

```diff
@@ -0,0 +1,205 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	capxv1 "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/external/github.com/nutanix-cloud-native/cluster-api-provider-nutanix/api/v1beta1"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pkg/webhook/preflight"
+)
+
+type failureDomainCheck struct {
+	failureDomainName string
+	namespace         string
+	field             string
+	kclient           ctrlclient.Client
+	nclient           client
+}
+
+func (fdc *failureDomainCheck) Name() string {
+	return "NutanixFailureDomains"
+}
+
+func newFailureDomainChecks(cd *checkDependencies) []preflight.Check {
+	checks := []preflight.Check{}
+
+	if cd.nclient == nil || cd.kclient == nil {
+		return checks
+	}
+
+	// For the failure domains configured for control-plane nodes
+	if cd.nutanixClusterConfigSpec != nil &&
+		cd.nutanixClusterConfigSpec.ControlPlane != nil &&
+		cd.nutanixClusterConfigSpec.ControlPlane.Nutanix != nil {
+		for _, fdName := range cd.nutanixClusterConfigSpec.ControlPlane.Nutanix.FailureDomains {
+			if fdName != "" {
+				checks = append(checks, &failureDomainCheck{
+					failureDomainName: fdName,
+					namespace:         cd.cluster.Namespace,
+					field: fmt.Sprintf(
+						"cluster.spec.topology.variables[.name=clusterConfig].value.controlPlane.nutanix.failureDomains[%s]",
+						fdName,
+					),
+					kclient: cd.kclient,
+					nclient: cd.nclient,
+				})
+			}
+		}
+	}
+
+	// For the failure domains configured for worker nodes
+	if cd.cluster != nil &&
+		cd.cluster.Spec.Topology != nil &&
+		cd.cluster.Spec.Topology.Workers != nil {
+		for i := range cd.cluster.Spec.Topology.Workers.MachineDeployments {
+			md := &cd.cluster.Spec.Topology.Workers.MachineDeployments[i]
+			if md.FailureDomain != nil && *md.FailureDomain != "" {
+				checks = append(checks, &failureDomainCheck{
+					failureDomainName: *md.FailureDomain,
+					namespace:         cd.cluster.Namespace,
+					field: fmt.Sprintf(
+						"cluster.spec.topology.workers.machineDeployments[.name=%s].failureDomain[%s]",
+						md.Name,
+						*md.FailureDomain,
+					),
+					kclient: cd.kclient,
+					nclient: cd.nclient,
+				})
+			}
+		}
+	}
+
+	return checks
+}
+
+func (fdc *failureDomainCheck) Run(ctx context.Context) preflight.CheckResult {
+	result := preflight.CheckResult{
+		Allowed: true,
+	}
+
+	// Fetch the referent failure domain object
+	fdObj := &capxv1.NutanixFailureDomain{}
+	fdKey := ctrlclient.ObjectKey{Name: fdc.failureDomainName, Namespace: fdc.namespace}
+	if err := fdc.kclient.Get(ctx, fdKey, fdObj); err != nil {
+		msg := ""
+		if errors.IsNotFound(err) {
+			msg = fmt.Sprintf("not found the failure domain object with name %q", fdc.failureDomainName)
+		} else {
+			msg = fmt.Sprintf("failed to fetch the failure domain object with name %q: %v", fdc.failureDomainName, err)
```

Comment:
I think we want to set 

CheckResult.InternalError to true here

Reply:
Failed to find the failureDomain object should not be an internalError. My understanding is only external (outside of k8s) errors such as the prism-api call errors should be internalError. cc: @thunderboltsid @dlipovetsky what do you think of this?

Reply:
> Failed to find the failureDomain object should not be an internalError. 

You are right. However, "failureDomain is not found" is handled on line 92.

Faiq is talking about line 95, where the error would be different, e.g. a DNS error, TCP error, etc.

Reply:
```suggestion
			msg = fmt.Sprintf(
                "Failed to get FailureDomain %q: %s. This is usually a temporary error. Please retry.",
                fdc.failureDomainName,
                err,
            )
```

Reply:
changed as suggested